File: app/src/main/java/com/cooper/wheellog/GarminConnectIQ.java
package com.cooper.wheellog;

import java.io.IOException;
import java.util.HashMap;
import java.util.List;
import java.util.Timer;
import java.util.TimerTask;

import android.app.Notification;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.os.Handler;
import android.os.IBinder;
import android.util.Log;
import android.widget.Toast;

import com.cooper.wheellog.utils.Constants;
import com.cooper.wheellog.utils.KingsongAdapter;

import com.cooper.wheellog.utils.NotificationUtil;
import com.cooper.wheellog.utils.SomeUtil;
import com.garmin.android.connectiq.ConnectIQ;
import com.garmin.android.connectiq.ConnectIQ.ConnectIQListener;
import com.garmin.android.connectiq.ConnectIQ.IQApplicationEventListener;
import com.garmin.android.connectiq.ConnectIQ.IQApplicationInfoListener;
import com.garmin.android.connectiq.ConnectIQ.IQConnectType;
import com.garmin.android.connectiq.ConnectIQ.IQDeviceEventListener;
import com.garmin.android.connectiq.ConnectIQ.IQMessageStatus;
import com.garmin.android.connectiq.ConnectIQ.IQSdkErrorStatus;
import com.garmin.android.connectiq.ConnectIQ.IQSendMessageListener;
import com.garmin.android.connectiq.IQApp;
import com.garmin.android.connectiq.IQDevice;
import com.garmin.android.connectiq.IQDevice.IQDeviceStatus;
import com.garmin.android.connectiq.exception.InvalidStateException;
import com.garmin.android.connectiq.exception.ServiceUnavailableException;
import com.garmin.monkeybrains.serialization.MonkeyHash;

import org.json.JSONException;
import org.json.JSONObject;

import fi.iki.elonen.NanoHTTPD;

public class GarminConnectIQ extends Service implements IQApplicationInfoListener, IQDeviceEventListener, IQApplicationEventListener, ConnectIQListener {
    public static final String TAG = GarminConnectIQ.class.getSimpleName();
    public static final String APP_ID = "df8bf0ab-1828-4037-a328-ee86d29d0501";
    private Notification mNotification;
    // This will require Garmin Connect V4.22
    // https://forums.garmin.com/developer/connect-iq/i/bug-reports/connect-version-4-20-broke-local-http-access
    public static final boolean FEATURE_FLAG_NANOHTTPD = true;

    public enum MessageType {
        EUC_DATA,
        PLAY_HORN,
        HTTP_READY,
    }

    public static final int MESSAGE_KEY_MSG_TYPE     = -2;
    public static final int MESSAGE_KEY_MSG_DATA     = -1;
    public static final int MESSAGE_KEY_SPEED        = 0;
    public static final int MESSAGE_KEY_BATTERY      = 1;
    public static final int MESSAGE_KEY_TEMPERATURE  = 2;
    public static final int MESSAGE_KEY_FAN_STATE    = 3;
    public static final int MESSAGE_KEY_BT_STATE     = 4;
    public static final int MESSAGE_KEY_VIBE_ALERT   = 5;
    public static final int MESSAGE_KEY_USE_MPH      = 6;
    public static final int MESSAGE_KEY_MAX_SPEED    = 7;
    public static final int MESSAGE_KEY_RIDE_TIME    = 8;
    public static final int MESSAGE_KEY_DISTANCE     = 9;
    public static final int MESSAGE_KEY_TOP_SPEED    = 10;
    public static final int MESSAGE_KEY_READY        = 11;
    public static final int MESSAGE_KEY_POWER        = 12;
    public static final int MESSAGE_KEY_ALARM1_SPEED = 13;
    public static final int MESSAGE_KEY_ALARM2_SPEED = 14;
    public static final int MESSAGE_KEY_ALARM3_SPEED = 15;

    public static final int MESSAGE_KEY_HTTP_PORT    = 99;

    int lastSpeed = 0;
    int lastBattery = 0;
    int lastTemperature = 0;
    int lastFanStatus = 0;
    int lastRideTime = 0;
    int lastDistance = 0;
    int lastTopSpeed = 0;
    boolean lastConnectionState = false;
    int lastPower = 0;

    private Timer keepAliveTimer;
    final Handler handler = new Handler(); // to call toast from the TimerTask

    private static GarminConnectIQ instance = null;

    private boolean mSdkReady;
    private ConnectIQ mConnectIQ;
    private List<IQDevice> mDevices;
    private IQDevice mDevice;
    private IQApp mMyApp;

    private GarminConnectIQWebServer mWebServer;

    public static boolean isInstanceCreated() {
        return instance != null;
    }

    public static GarminConnectIQ instance() {
        return instance;
    }

    @Override
    public IBinder onBind(Intent intent) {
        Log.i(TAG, "onBind");
        return null;
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        Log.d(TAG,"onStartCommand");
        super.onStartCommand(intent, flags, startId);

        instance = this;

        // Setup Connect IQ
        mMyApp = new IQApp(APP_ID);
        mConnectIQ = ConnectIQ.getInstance(this, IQConnectType.WIRELESS);
        mConnectIQ.initialize(this, true, this);
        startForeground(Constants.MAIN_NOTIFICATION_ID, NotificationUtil.getNotification());
        return START_STICKY;
    }

    @Override
    public void onDestroy() {
        Log.d(TAG,"onDestroy");
        super.onDestroy();

        cancelRefreshTimer();

        try {
            mConnectIQ.unregisterAllForEvents();
            mConnectIQ.shutdown(this);
        } catch (InvalidStateException e) {
            // This is usually because the SDK was already shut down
            // so no worries.
        }

        stopWebServer();
        stopForeground(false);
        instance = null;
    }

    // General METHODS
    private void populateDeviceList() {
        Log.d(TAG,"populateDeviceList");

        try {
            mDevices = mConnectIQ.getKnownDevices();

            if (mDevices != null && !mDevices.isEmpty()) {
                mDevice = mDevices.get(0);
                registerWithDevice();
            }

        } catch (InvalidStateException e) {
            // This generally means you forgot to call initialize(), but since
            // we are in the callback for initialize(), this should never happen
        } catch (ServiceUnavailableException e) {
            // This will happen if for some reason your app was not able to connect
            // to the ConnectIQ service running within Garmin Connect Mobile.  This
            // could be because Garmin Connect Mobile is not installed or needs to
            // be upgraded.
            Toast.makeText(this, R.string.garmin_connectiq_service_unavailable_message, Toast.LENGTH_LONG).show();
        }
    }

    private void registerWithDevice() {
        Log.d(TAG,"registerWithDevice");

        if (mDevice != null && mSdkReady) {
            // Register for device status updates
            try {
                mConnectIQ.registerForDeviceEvents(mDevice, this);
            } catch (InvalidStateException e) {
                Log.wtf(TAG, "InvalidStateException:  We should not be here!");
            }

            // Register for application status updates
            try {
                mConnectIQ.getApplicationInfo(APP_ID, mDevice, this);
            } catch (InvalidStateException e1) {
                Log.d(TAG, "e1: " + e1.getMessage());
            } catch (ServiceUnavailableException e1) {
                Log.d(TAG, "e2: " + e1.getMessage());
            }

            // Register to receive messages from the device
            try {
                mConnectIQ.registerForAppEvents(mDevice, mMyApp, this);
            } catch (InvalidStateException e) {
                Toast.makeText(this, "ConnectIQ is not in a valid state", Toast.LENGTH_LONG).show();
            }
        }
    }

    private void unregisterWithDevice() {
        Log.d(TAG,"unregisterWithDevice");

        if (mDevice != null && mSdkReady) {
            // It is a good idea to unregister everything and shut things down to
            // release resources and prevent unwanted callbacks.
            try {
                mConnectIQ.unregisterForDeviceEvents(mDevice);

                if (mMyApp != null) {
                    mConnectIQ.unregisterForApplicationEvents(mDevice, mMyApp);
                }
            } catch (InvalidStateException e) {
            }
        }
    }

    private void cancelRefreshTimer() {
        if (keepAliveTimer != null) {
            keepAliveTimer.cancel();
            keepAliveTimer = null;
        }
    }

    private void startRefreshTimer() {
        TimerTask timerTask = new TimerTask() {
            @Override
            public void run() {
                handler.post(new Runnable() {
                    public void run() {
                        refreshData();
                    }
                });
            }
        };

        keepAliveTimer = new Timer();
        keepAliveTimer.scheduleAtFixedRate(timerTask, 0, 1500); // 1.5cs
    }

    private void refreshData() {
        if (WheelData.getInstance() == null)
            return;

        try {
            HashMap<Object, Object> data = new HashMap<Object, Object>();

            lastSpeed = WheelData.getInstance().getSpeed() / 10;
            data.put(MESSAGE_KEY_SPEED, lastSpeed);

            lastBattery = WheelData.getInstance().getBatteryLevel();
            data.put(MESSAGE_KEY_BATTERY, lastBattery);

            lastTemperature = WheelData.getInstance().getTemperature();
            data.put(MESSAGE_KEY_TEMPERATURE, lastTemperature);

            lastFanStatus = WheelData.getInstance().getFanStatus();
            data.put(MESSAGE_KEY_FAN_STATE, lastFanStatus);

            lastConnectionState = WheelData.getInstance().isConnected();
            data.put(MESSAGE_KEY_BT_STATE, lastConnectionState);

            data.put(MESSAGE_KEY_VIBE_ALERT, false);
            data.put(MESSAGE_KEY_USE_MPH, WheelLog.AppConfig.getUseMph());
            data.put(MESSAGE_KEY_MAX_SPEED, WheelLog.AppConfig.getMaxSpeed());

            lastRideTime = WheelData.getInstance().getRideTime();
            data.put(MESSAGE_KEY_RIDE_TIME, lastRideTime);

            lastDistance = WheelData.getInstance().getDistance();
            data.put(MESSAGE_KEY_DISTANCE, lastDistance/100);

            lastTopSpeed = WheelData.getInstance().getTopSpeed();
            data.put(MESSAGE_KEY_TOP_SPEED, lastTopSpeed/10);

            lastPower = (int)WheelData.getInstance().getPowerDouble();
            data.put(MESSAGE_KEY_POWER, lastPower);

            HashMap<Object, Object> message = new HashMap<Object, Object>();
            message.put(MESSAGE_KEY_MSG_TYPE, MessageType.EUC_DATA.ordinal());
            message.put(MESSAGE_KEY_MSG_DATA, new MonkeyHash(data));

            try {
                mConnectIQ.sendMessage(mDevice, mMyApp, message, new IQSendMessageListener() {

                    @Override
                    public void onMessageStatus(IQDevice device, IQApp app, IQMessageStatus status) {
                        Log.d(TAG, "message status: " + status.name());

                        if (status.name() != "SUCCESS")
                            Toast.makeText(GarminConnectIQ.this, status.name(), Toast.LENGTH_LONG).show();
                    }

                });
            } catch (InvalidStateException e) {
                Log.e(TAG, "ConnectIQ is not in a valid state");
                Toast.makeText(this, "ConnectIQ is not in a valid state", Toast.LENGTH_LONG).show();
            } catch (ServiceUnavailableException e) {
                Log.e(TAG, "ConnectIQ service is unavailable.   Is Garmin Connect Mobile installed and running?");
                Toast.makeText(this, "ConnectIQ service is unavailable.   Is Garmin Connect Mobile installed and running?", Toast.LENGTH_LONG).show();
            }
        } catch (Exception ex) {
            Log.e(TAG, "refreshData", ex);
            Toast.makeText(this, "Error refreshing data", Toast.LENGTH_SHORT).show();
        }
    }

    // IQApplicationInfoListener METHODS
    @Override
    public void onApplicationInfoReceived(IQApp app) {
        Log.d(TAG,"onApplicationInfoReceived");
        Log.d(TAG, app.toString());
    }

    @Override
    public void onApplicationNotInstalled(String arg0) {
        Log.d(TAG,"onApplicationNotInstalled");

        // The WheelLog app is not installed on the device so we have
        // to let the user know to install it.
        cancelRefreshTimer(); // no point in sending data...

        Toast.makeText(this, R.string.garmin_connectiq_missing_app_message, Toast.LENGTH_LONG).show();
        try {
            mConnectIQ.openStore(APP_ID);
        } catch (InvalidStateException e) {

        } catch (ServiceUnavailableException e) {

        }
    }

    // IQDeviceEventListener METHODS
    @Override
    public void onDeviceStatusChanged(IQDevice device, IQDeviceStatus status) {
        Log.d(TAG,"onDeviceStatusChanged");
        Log.d(TAG, "status is:" + status.name());

        switch(status.name()) {
            case "CONNECTED":
                // Disabled the push method for now until a dev from garmin can shed some light on the
                // intermittent FAILURE_DURING_TRANSFER that we have seen. This is documented here:
                // https://forums.garmin.com/developer/connect-iq/f/legacy-bug-reports/5144/failure_during_transfer-issue-again-now-using-comm-sample
                if (!FEATURE_FLAG_NANOHTTPD) {
                    startRefreshTimer();
                }

                // As a workaround, start a nanohttpd server that will listen for data requests from the watch. This is
                // also documented on the link above and is apparently a good workaround for the meantime. In our implementation
                // we instanciate the httpd server on an ephemeral port and send a message to the watch to tell it on which port
                // it can request its data.
                if (FEATURE_FLAG_NANOHTTPD) {
                    startWebServer();
                }
                break;
            case "NOT_PAIRED":
            case "NOT_CONNECTED":
            case "UNKNOWN":
                cancelRefreshTimer();
                stopWebServer();
        }
    }

    // IQApplicationEventListener
    @Override
    public void onMessageReceived(IQDevice device, IQApp app, List<Object> message, IQMessageStatus status) {
        Log.d(TAG,"onMessageReceived");

        // We know from our widget that it will only ever send us strings, but in case
        // we get something else, we are simply going to do a toString() on each object in the
        // message list.
        StringBuilder builder = new StringBuilder();

        if (message.size() > 0) {
            for (Object o : message) {
                if (o == null) {
                    builder.append("<null> received");
                } else if (o instanceof HashMap) {
                    try {
                        HashMap msg = (HashMap)o;
                        int msgType = (int)msg.get(MESSAGE_KEY_MSG_TYPE);
                        if (msgType == MessageType.PLAY_HORN.ordinal()) {
                            playHorn();
                        }
                        builder = null;
                    } catch (Exception ex) {
                        builder.append("MonkeyHash received:\n\n");
                        builder.append(o.toString());
                    }
                } else {
                    builder.append(o.toString());
                    builder.append("\r\n");
                }
            }
        } else {
            builder.append("Received an empty message from the ConnectIQ application");
        }

        if (builder != null) {
            Toast.makeText(getApplicationContext(), builder.toString(), Toast.LENGTH_SHORT).show();
        }
    }

    // ConnectIQListener METHODS
    @Override
    public void onInitializeError(IQSdkErrorStatus errStatus) {
        Log.d(TAG,"sdk initialization error");
        mSdkReady = false;
    }

    @Override
    public void onSdkReady() {
        Log.d(TAG,"sdk is ready");
        mSdkReady = true;
        populateDeviceList();
    }

    @Override
    public void onSdkShutDown() {
        Log.d(TAG,"sdk shut down");
        mSdkReady = false;
    }

    public void playHorn() {
        Context context = getApplicationContext();

        int horn_mode = WheelLog.AppConfig.getHornMode();
        if (horn_mode == 1 && WheelData.getInstance().getWheelType() == Constants.WHEEL_TYPE.KINGSONG) {
            KingsongAdapter.getInstance().horn();
        } else if (horn_mode == 2) {
            SomeUtil.playSound(context, R.raw.bicycle_bell);
        }
    }

    public void startWebServer() {
        Log.d(TAG,"startWebServer");

        if (mWebServer != null)
            return;

        try {
            mWebServer = new GarminConnectIQWebServer();
            Log.d(TAG, "port is:" + mWebServer.getListeningPort());

            HashMap<Object, Object> data = new HashMap<Object, Object>();
            data.put(MESSAGE_KEY_HTTP_PORT, mWebServer.getListeningPort());

            HashMap<Object, Object> message = new HashMap<Object, Object>();
            message.put(MESSAGE_KEY_MSG_TYPE, MessageType.HTTP_READY.ordinal());
            message.put(MESSAGE_KEY_MSG_DATA, new MonkeyHash(data));

            try {
                mConnectIQ.sendMessage(mDevice, mMyApp, message, new IQSendMessageListener() {

                    @Override
                    public void onMessageStatus(IQDevice device, IQApp app, IQMessageStatus status) {
                        Log.d(TAG, "message status: " + status.name());

                        if (status.name() != "SUCCESS")
                            Toast.makeText(GarminConnectIQ.this, status.name(), Toast.LENGTH_LONG).show();
                    }

                });
            } catch (InvalidStateException e) {
                Log.e(TAG, "ConnectIQ is not in a valid state");
                Toast.makeText(this, "ConnectIQ is not in a valid state", Toast.LENGTH_LONG).show();
            } catch (ServiceUnavailableException e) {
                Log.e(TAG, "ConnectIQ service is unavailable.   Is Garmin Connect Mobile installed and running?");
                Toast.makeText(this, "ConnectIQ service is unavailable.   Is Garmin Connect Mobile installed and running?", Toast.LENGTH_LONG).show();
            }

        } catch (IOException e) {
        }
    }

    public void stopWebServer() {
        Log.d(TAG,"stopWebServer");
        if (mWebServer != null) {
            mWebServer.stop();
            mWebServer = null;
        }
    }
}

class GarminConnectIQWebServer extends NanoHTTPD {
    public GarminConnectIQWebServer()throws IOException {
        super("127.0.0.1", 0); // 0 to automatically find an available ephemeral port
        start(NanoHTTPD.SOCKET_READ_TIMEOUT, false);
    }

    @Override
    public Response serve(IHTTPSession session)
    {
        if (session.getMethod() == Method.GET && session.getUri().equals("/data")) {
            return handleData();
        }

        return newFixedLengthResponse(Response.Status.NOT_FOUND, NanoHTTPD.MIME_PLAINTEXT, "Error 404, file not found.");
    }

    private Response handleData() {
        Log.d("GarminConnectIQWebSe...","handleData");
        JSONObject data = new JSONObject();

        try {
            data.put("" + GarminConnectIQ.MESSAGE_KEY_SPEED, WheelData.getInstance().getSpeed() / 10); // Convert to km/h
            data.put("" + GarminConnectIQ.MESSAGE_KEY_BATTERY, WheelData.getInstance().getBatteryLevel());
            data.put("" + GarminConnectIQ.MESSAGE_KEY_TEMPERATURE, WheelData.getInstance().getTemperature());
            data.put("" + GarminConnectIQ.MESSAGE_KEY_FAN_STATE, WheelData.getInstance().getFanStatus());
            data.put("" + GarminConnectIQ.MESSAGE_KEY_BT_STATE, WheelData.getInstance().isConnected());
            data.put("" + GarminConnectIQ.MESSAGE_KEY_VIBE_ALERT, false);
            data.put("" + GarminConnectIQ.MESSAGE_KEY_USE_MPH, WheelLog.AppConfig.getUseMph());
            data.put("" + GarminConnectIQ.MESSAGE_KEY_MAX_SPEED, WheelLog.AppConfig.getMaxSpeed());
            data.put("" + GarminConnectIQ.MESSAGE_KEY_RIDE_TIME, WheelData.getInstance().getRideTime());
            data.put("" + GarminConnectIQ.MESSAGE_KEY_DISTANCE, WheelData.getInstance().getDistance() / 100);
            data.put("" + GarminConnectIQ.MESSAGE_KEY_TOP_SPEED, WheelData.getInstance().getTopSpeed() / 10);
            data.put("" + GarminConnectIQ.MESSAGE_KEY_POWER, (int) WheelData.getInstance().getPowerDouble());
            if (WheelData.getInstance().getWheelType() == Constants.WHEEL_TYPE.KINGSONG) {
                data.put("" + GarminConnectIQ.MESSAGE_KEY_ALARM1_SPEED, WheelLog.AppConfig.getWheelKsAlarm1());
                data.put("" + GarminConnectIQ.MESSAGE_KEY_ALARM2_SPEED, WheelLog.AppConfig.getWheelKsAlarm2());
                data.put("" + GarminConnectIQ.MESSAGE_KEY_ALARM3_SPEED, WheelLog.AppConfig.getWheelKsAlarm3());
            }

            JSONObject message = new JSONObject();
            message.put("" + GarminConnectIQ.MESSAGE_KEY_MSG_TYPE, GarminConnectIQ.MessageType.EUC_DATA.ordinal());
            message.put("" + GarminConnectIQ.MESSAGE_KEY_MSG_DATA, data);

            return newFixedLengthResponse(Response.Status.OK, "application/json", message.toString());
        } catch (JSONException e) {
            return newFixedLengthResponse(Response.Status.SERVICE_UNAVAILABLE, "text/plain", "");
        }
    }
}

File: app/src/main/java/com/cooper/wheellog/GarminConnectIQ.kt
package com.cooper.wheellog

import android.app.Service
import android.content.Context
import android.content.Intent
import android.os.IBinder
import android.widget.Toast
import com.cooper.wheellog.utils.Alarms
import com.cooper.wheellog.utils.AudioUtil
import com.cooper.wheellog.utils.Constants
import androidx.core.math.MathUtils
import com.cooper.wheellog.utils.*
import com.cooper.wheellog.utils.SomeUtil.Companion.playBeep
import com.garmin.android.connectiq.ConnectIQ.*
import com.garmin.android.connectiq.IQApp
import com.garmin.android.connectiq.IQDevice
import com.garmin.android.connectiq.IQDevice.IQDeviceStatus
import com.garmin.android.connectiq.exception.InvalidStateException
import com.garmin.android.connectiq.exception.ServiceUnavailableException
import fi.iki.elonen.NanoHTTPD
import org.json.JSONObject
import timber.log.Timber
import java.io.IOException
import java.util.*

class GarminConnectIQ : Service(), IQApplicationInfoListener, IQDeviceEventListener, IQApplicationEventListener, ConnectIQListener {
    private var keepAliveTimer: Timer? = null
    private var mSdkReady = false
    private var mConnectIQ = getInstance(this, IQConnectType.WIRELESS)
    private var mDevice: IQDevice? = null
    private var mApp: IQApp? = null
    private var mWebServer: GarminConnectIQWebServer? = null
    private var useBeta = WheelLog.AppConfig.useGarminBetaCompanion

    override fun onBind(intent: Intent): IBinder? {
        Timber.i("onBind")
        return null
    }

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        Timber.d("onStartCommand")
        super.onStartCommand(intent, flags, startId)
        instance = this

        // Setup Connect IQ
        mApp = if (useBeta) {
            IQApp(BETA_APP_ID)
        } else {
            IQApp(STABLE_APP_ID)
        }
        mConnectIQ.initialize(this, false, this)
        startForeground(Constants.MAIN_NOTIFICATION_ID, WheelLog.Notifications.notification)
        return START_STICKY
    }

    override fun onDestroy() {
        Timber.d("onDestroy")
        super.onDestroy()
        cancelRefreshTimer()
        try {
            mConnectIQ.unregisterAllForEvents()
            mConnectIQ.shutdown(this)
        } catch (e: InvalidStateException) {
            // This is usually because the SDK was already shut down
            // so no worries.
        }
        stopWebServer()
        unregisterWithDevice()
        stopForeground(false)
        instance = null
    }

    // General METHODS
    private fun populateDeviceList() {
        Timber.d("populateDeviceList")
        try {
            val mDevices = mConnectIQ.knownDevices
            if (mDevices != null && mDevices.isNotEmpty()) {
                mDevice = mDevices[0]
                registerWithDevice()
            }
        } catch (e: InvalidStateException) {
            // This generally means you forgot to call initialize(), but since
            // we are in the callback for initialize(), this should never happen
        } catch (e: ServiceUnavailableException) {
            // This will happen if for some reason your app was not able to connect
            // to the ConnectIQ service running within Garmin Connect Mobile.  This
            // could be because Garmin Connect Mobile is not installed or needs to
            // be upgraded.
            Toast.makeText(this, R.string.garmin_connectiq_service_unavailable_message, Toast.LENGTH_LONG).show()
        }
    }

    private fun registerWithDevice() {
        Timber.d("registerWithDevice")
        if (mDevice != null && mSdkReady) {
            // Register for device status updates
            try {
                mConnectIQ.registerForDeviceEvents(mDevice, this)
            } catch (e: InvalidStateException) {
                Timber.wtf("InvalidStateException:  We should not be here!")
            }

            // Register for application status updates
            try {
                if (useBeta) {
                    mConnectIQ.getApplicationInfo(BETA_APP_ID, mDevice, this)
                } else {
                    mConnectIQ.getApplicationInfo(STABLE_APP_ID, mDevice, this)
                }
            } catch (e1: InvalidStateException) {
                Timber.d("e1: ${e1.message}")
            } catch (e1: ServiceUnavailableException) {
                Timber.d("e2: ${e1.message}")
            }

            // Register to receive messages from the device
            try {
                mConnectIQ.registerForAppEvents(mDevice, mApp, this)
            } catch (e: InvalidStateException) {
                Toast.makeText(this, "ConnectIQ is not in a valid state", Toast.LENGTH_LONG).show()
            }
        }
    }

    private fun unregisterWithDevice() {
        Timber.d("unregisterWithDevice")
        if (mDevice != null && mSdkReady) {
            // It is a good idea to unregister everything and shut things down to
            // release resources and prevent unwanted callbacks.
            try {
                mConnectIQ.unregisterForDeviceEvents(mDevice)
                if (mApp != null) {
                    mConnectIQ.unregisterForApplicationEvents(mDevice, mApp)
                }
            } catch (ignored: InvalidStateException) {

            }
        }
    }

    private fun cancelRefreshTimer() {
        keepAliveTimer?.cancel()
        keepAliveTimer = null
    }

    // IQApplicationInfoListener METHODS
    override fun onApplicationInfoReceived(app: IQApp) {
        Timber.d("onApplicationInfoReceived")
        Timber.d(app.toString())
    }

    override fun onApplicationNotInstalled(arg0: String) {
        Timber.d("onApplicationNotInstalled")

        // The WheelLog app is not installed on the device so we have
        // to let the user know to install it.
        cancelRefreshTimer() // no point in sending data...
        Toast.makeText(this, R.string.garmin_connectiq_missing_app_message, Toast.LENGTH_LONG).show()
        try {
            mConnectIQ.openStore("35719a02-8a5d-46bc-b474-f26c54c4e045")
        } catch (ignored: InvalidStateException) {
        } catch (ignored: ServiceUnavailableException) {
        }
    }

    // IQDeviceEventListener METHODS
    override fun onDeviceStatusChanged(device: IQDevice, status: IQDeviceStatus) {
        Timber.d("onDeviceStatusChanged")
        Timber.d("status is: ${status.name}")
        if (status.name == "CONNECTED") {
            startWebServer()
        } else {
            cancelRefreshTimer()
            stopWebServer()
        }
    }

    // IQApplicationEventListener
    override fun onMessageReceived(device: IQDevice, app: IQApp, message: List<Any>, status: IQMessageStatus) {
        Timber.d("onMessageReceived")
        // This thing won't do anything, because every data transmit is done through a web server
    }

    // ConnectIQListener METHODS
    override fun onInitializeError(errStatus: IQSdkErrorStatus) {
        Timber.d("sdk initialization error")
        mSdkReady = false
    }

    override fun onSdkReady() {
        Timber.d("sdk is ready")
        mSdkReady = true
        populateDeviceList()
    }

    override fun onSdkShutDown() {
        Timber.d("sdk shut down")
        mSdkReady = false
    }

    private fun startWebServer() {
        Timber.d("startWebServer")
        if (mWebServer != null) return
        try {
            mWebServer = GarminConnectIQWebServer(applicationContext)
            Timber.d("port is: ${mWebServer!!.listeningPort}")
            try {
                mConnectIQ.sendMessage(mDevice, mApp, mWebServer!!.listeningPort) { _: IQDevice?, _: IQApp?, status: IQMessageStatus ->
                    Timber.d("message status: ${status.name}")
                    if (status.name !== "SUCCESS") Toast.makeText(this@GarminConnectIQ, status.name, Toast.LENGTH_LONG).show()
                }
            } catch (e: InvalidStateException) {
                Timber.e("ConnectIQ is not in a valid state")
                Toast.makeText(this, "ConnectIQ is not in a valid state", Toast.LENGTH_LONG).show()
            } catch (e: ServiceUnavailableException) {
                Timber.e("ConnectIQ service is unavailable.   Is Garmin Connect Mobile installed and running?")
                Toast.makeText(this, "ConnectIQ service is unavailable.   Is Garmin Connect Mobile installed and running?", Toast.LENGTH_LONG).show()
            }
        } catch (e: IOException) {
            Timber.e(e, "IOException happened while starting ConnectIQ web server")
        }
    }

    private fun stopWebServer() {
        Timber.d("stopWebServer")
        mWebServer?.stop()
        mWebServer = null
    }

    companion object {
        const val STABLE_APP_ID = "487e6172-972c-4f93-a4db-26fd689f935a"
        const val BETA_APP_ID = "433c30dc-f316-4d11-a16e-de153d297705"

        private var instance: GarminConnectIQ? = null
        val isInstanceCreated: Boolean
            get() = instance != null
    }
}

internal class GarminConnectIQWebServer(context: Context) : NanoHTTPD("127.0.0.1", 0) {
    private var applicationContext: Context

    init {
        start(SOCKET_READ_TIMEOUT, false)
        applicationContext = context
    }

    private fun playHorn() {
        playBeep(WheelLog.AppConfig.hornMode == 1, false)
    }

    private val speedStr
        get() = run {
            val speed = if (!WheelLog.AppConfig.useMph) {
                WheelData.getInstance().speedDouble
            } else {
                MathsUtil.kmToMiles(WheelData.getInstance().speedDouble)
            }
            if (speed.toString().length > 3) {
                ((speed * 10).toInt().toFloat() / 10).toString()
            } else speed.toString()
        }

    private val topSpeed
        get() = if (!WheelLog.AppConfig.useMph) {
            WheelData.getInstance().topSpeed
        } else {
            MathsUtil.kmToMiles(WheelData.getInstance().topSpeed.toDouble()).toInt()
        }

    private val temperature
        get() = if (!WheelLog.AppConfig.useMph) {
            WheelData.getInstance().temperature
        } else {
            MathsUtil.celsiusToFahrenheit(WheelData.getInstance().temperature.toDouble()).toInt()
        }

    private val avgSpeed
        get() = if (!WheelLog.AppConfig.useMph) {
            WheelData.getInstance().averageSpeedDouble.toInt()
        } else {
            MathsUtil.kmToMiles(WheelData.getInstance().averageSpeedDouble).toInt()
        }

    private val avgRidingSpeed
        get() = if (!WheelLog.AppConfig.useMph) {
            WheelData.getInstance().averageRidingSpeedDouble.toInt()
        } else {
            MathsUtil.kmToMiles(WheelData.getInstance().averageRidingSpeedDouble).toInt()
        }

    override fun serve(session: IHTTPSession): Response {
        val wd = WheelData.getInstance()
        val ac = WheelLog.AppConfig
      
        return when (session.method) {
            Method.GET -> {
                when (session.uri) {
                    "/data/main" -> {
                        val message = JSONObject()
                        message.put("speed", speedStr)
                        message.put("topSpeed", ((topSpeed / 10).toFloat() / 10).toString())
                        message.put("speedLimit", ac.maxSpeed)
                        message.put("useMph", ac.useMph)
                        message.put("battery", wd.batteryLevel)
                        message.put("temp", temperature)
                        message.put("pwm", String.format("%02.0f", wd.calculatedPwm))
                        message.put("maxPwm", String.format("%02.0f", wd.maxPwm))
                        message.put("connectedToWheel", wd.isConnected)
                        message.put("wheelModel", wd.model)

                        return newFixedLengthResponse(Response.Status.OK, "application/json", message.toString()) // Send data
                    }
                    "/data/details" -> {
                        val message = JSONObject()
                        message.put("useMph", ac.useMph)
                        message.put("avgRidingSpeed", avgRidingSpeed)
                        message.put("avgSpeed", avgSpeed)
                        message.put("topSpeed", ((topSpeed / 10).toFloat() / 10).toString())
                        message.put("voltage", wd.voltageDouble.toString())
                        message.put("maxVoltage", wd.maxVoltageForWheel.toString())
                        message.put("battery", wd.batteryLevel)
                        message.put("ridingTime", wd.ridingTimeString)
                        message.put("distance", wd.distance)
                        message.put("pwm", String.format("%02.0f", wd.calculatedPwm))
                        message.put("maxPwm", String.format("%02.0f", wd.maxPwm))
                        message.put("torque", wd.torque)
                        message.put("power", wd.powerDouble)
                        message.put("maxPower", wd.maxPowerDouble)

                        message.put("connectedToWheel", wd.isConnected)

                        return newFixedLengthResponse(Response.Status.OK, "application/json", message.toString()) // Send data
                    }
                    "/data/alarms" -> {
                        val message = "${Alarms.alarm}"
                        newFixedLengthResponse(Response.Status.OK, "application/json", message) // Send data
                    }
                    else -> {
                        Timber.i("404 Wrong endpoint")
                        newFixedLengthResponse(Response.Status.NOT_FOUND, MIME_PLAINTEXT, "404: File not found")
                    }
                }
            }
            Method.POST -> {
                when (session.uri) {
                    "/actions/triggerHorn" -> {
                        playHorn()
                        newFixedLengthResponse(Response.Status.OK, MIME_PLAINTEXT, "Executed!") // Send data
                    }
                    "/actions/frontLight/enable" -> {
                        wd.updateLight(true)
                        newFixedLengthResponse(Response.Status.OK, MIME_PLAINTEXT, "Executed!") // Send data
                    }
                    "/actions/frontLight/disable" -> {
                        newFixedLengthResponse(Response.Status.OK, MIME_PLAINTEXT, "Executed!") // Send data
                    }
                    else -> newFixedLengthResponse(Response.Status.NOT_FOUND, MIME_PLAINTEXT, "Error 404, action not found.")
                }
            }
            else -> {
                Timber.i("404 Wrong method")
                newFixedLengthResponse(Response.Status.NOT_FOUND, MIME_PLAINTEXT, "404: File not found")
            }
        }
    }
}